src/gc/gc.h

```diff
@@ -352,19 +197,7 @@ void record_changed_seg (uint8_t* start, uint8_t* end,
 void record_global_mechanism (int mech_index);
 #endif //GC_CONFIG_DRIVEN
 
-//constants for the flags parameter to the gc call back
-
-#define GC_CALL_INTERIOR            0x1
-#define GC_CALL_PINNED              0x2
-#define GC_CALL_CHECK_APP_DOMAIN    0x4
-
-//flags for GCHeap::Alloc(...)
-#define GC_ALLOC_FINALIZE 0x1
-#define GC_ALLOC_CONTAINS_REF 0x2
-#define GC_ALLOC_ALIGN8_BIAS 0x4
-#define GC_ALLOC_ALIGN8 0x8
-
-class GCHeap {
+class GCHeap : public IGCHeap {
```

Comment:
@jkotas (and @janvorli since it's related to the VM managing the GC instance)

I took a look at the role of `GCHeap` and `g_pGCHeap` within the GC and found a couple of places that 1) the GC assumes that `g_pGCHeap` exists and 2) the GC calls into itself by invoking `GCHeap` methods on `g_pGCHeap`, e.g:
```c++
void fix_alloc_context(alloc_context* acontext, void* param)
{
    fix_alloc_context_args* args = (fix_alloc_context_args*)param;
    GCHeap::GetGCHeap()->FixAllocContext(acontext, FALSE, (void*)(size_t)(args->for_gc_p), args->heap);
}
```

```c++
void gc_heap::descr_generations_to_profiler (gen_walk_fn fn, void *context)
{
#ifdef MULTIPLE_HEAPS
    int n_heaps = GCHeap::GetGCHeap()->GetNumberOfHeaps ();
    for (int i = 0; i < n_heaps; i++)
    {
[...]
```

From what I can see, if we are aiming to move the management of the singular `IGCHeap` instance to the VM, we will need to fix the cases within the GC where we are relying on the global heap instance.

An additional side effect of the GC calling into itself in this manner is that it forces these calls to be on the `IGCHeap` interface. If we removed `GCHeap` and made `SVR::GCHeap` and `WKS::GCHeap` derive from `IGCHeap` directly, these internal self-calls must go through the interface and thus, the calls that they make must also be on the interface. 

I'd imagine that all of these are fixable by referencing internals instead of calling into the API, but I'd have to look at each one of these calls to be sure of that.

I suppose another option would to add an API to the `GCToEEInterface` that retrieves `g_pGCHeap`, since the VM would be responsible for it.

Do you all have any thoughts on this?